### PR TITLE
HTTP API: Allow binary downloads and headers

### PR DESCRIPTION
**Pull request by SmallJoker**
This patch allows to download binary files.
Needed for https://github.com/minetest-mods/skinsdb/pull/27

How to test:
1) Request the HTTP functions
2) Download s binary file
3) Write it to the disk

Previously the contents were only copied to the first NULL character (C-strings).

### CODE REVIEW

**Comment by sfan5** [comment]
This would be a good candidate to add to `minetest.features` so mods can check whether binary requests are supported.

**Comment by thomasrudin** [comment]
Does this also support binary uploads?

**Comment by SmallJoker** [comment]
@thomasrudin Yes, after this PR you can pass them in POST fields. I don't think it would be a good idea to allow that in the URL (i.e. GET fields) as well. I did not test this functionality but `HTTPFetchOngoing::HTTPFetchOngoing` clearly passes cURL the length of the C-string which then is used instead of the null termination.

@sfan5 Thanks for the suggestion. Added.

File: src/script/common/c_converter.cpp [hunk 1]
```diff
@@ -540,9 +540,9 @@ v3s16 getv3s16field_default(lua_State *L, int table,
 }
 
 void setstringfield(lua_State *L, int table,
-		const char *fieldname, const char *value)
+		const char *fieldname, const std::string &value)
```
Comment (sfan5) [inline src/script/common/c_converter.cpp]: does this not introduce some/many(?) unneeded conversions to std::string and back?
how often is setstringfield used with `char*`?

Comment (SmallJoker) [reply src/script/common/c_converter.cpp]: It's used exactly 1 times: `l_http.cpp`

**Pull request merged**